### PR TITLE
Add elementwise_affine for layer_norm_op

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary: Add elementwise_affine for layer_norm_op

Differential Revision: D15075454

### CODE REVIEW

File: caffe2/operators/layer_norm_op.cc [hunk 1]
```diff
@@ -174,10 +191,16 @@ to the end.)
...
         " dividing out by that value. This prevents division by zero.")
+    .Arg(
+        "elementwise_affine",
+        "(bool) default to False; If true, this op will do affine "
+        "transformation after normalization.")
     .Input(
         0,
         "input",
         "Input tensor which layer normalization will be applied to")
+    .Input(1, "gamma", "scale tensor for elementwise_affine")
```
Comment (houseroad) [inline caffe2/operators/layer_norm_op.cc]: specify expected shape?

File: caffe2/operators/layer_norm_op.cc [hunk 2]
```diff
@@ -174,10 +191,16 @@ to the end.)
         "epsilon",
         "(float) default to 0.001. Small value to be added to the stdev when"
         " dividing out by that value. This prevents division by zero.")
+    .Arg(
+        "elementwise_affine",
```
Comment (houseroad) [inline caffe2/operators/layer_norm_op.cc]: This seems redundant, since we can determine whether we should do the elementwise_affine based on the input size.

File: caffe2/python/operator_test/layer_norm_op_test.py [hunk 3]
```diff
@@ -6,28 +6,35 @@
...
+
 import numpy as np
 import os
 import unittest
 from functools import partial
 import torch
 
 
-def _layer_norm_ref(axis, epsilon, X):
+def _layer_norm_ref1(axis, epsilon, X):
```
Comment (houseroad) [inline caffe2/python/operator_test/layer_norm_op_test.py]: nit: _layer_norm_ref

File: caffe2/python/operator_test/layer_norm_op_test.py [hunk 4]
```diff
@@ -6,28 +6,35 @@
...
+    Y = (reshaped - mean) / (std)
+    Y = np.reshape(Y, X.shape)
     mean = np.reshape(mean, X.shape[:axis] + (1,))
-    stdev = np.reshape(stdev, X.shape[:axis] + (1,))
-    return [norm, mean, stdev]
+    std = np.reshape(std, X.shape[:axis] + (1,))
+    return (Y, mean, std)
+
+
+def _layer_norm_ref2(axis, epsilon, X, gamma, beta):
```
Comment (houseroad) [inline caffe2/python/operator_test/layer_norm_op_test.py]: nit: _layer_norm_with_affine_ref?

File: caffe2/utils/math/reduce.cc [hunk 5]
```diff
@@ -663,6 +663,7 @@ DELEGATE_REDUCE_FUNCTION(double, ReduceL2, 0.0, true)
     MomentsImpl<T>(ndim, X_dims, Y_dims, X, mean, var, context); \
   }
 CAFFE2_SPECIALIZED_MOMENTS(float)
+CAFFE2_SPECIALIZED_MOMENTS(double)
```
Comment (houseroad) [inline caffe2/utils/math/reduce.cc]: Is this necessary?

**Review by houseroad — COMMENTED** [review]
Overall, looks good. Please check inline comments.

File: caffe2/core/operator_c10wrapper.h [hunk 6]
```diff
@@ -103,15 +108,14 @@ class C10OperatorWrapper final : public Operator<Context> {
...
-
-      AT_ASSERTM(
-          input_tensor_index == InputSize(),
-          "Error in caffe2->c10 wrapper: Number of caffe2 operator inputs (",
-          InputSize(),
-          ") doesn't match number of tensor arguments (",
-          input_tensor_index,
-          ") in the c10 operator schema.");
     }
+    AT_ASSERTM(
```
Comment (smessmer) [inline caffe2/core/operator_c10wrapper.h]: thanks. awesome catch.

File: caffe2/operators/layer_norm_op.cc (hunk 2 above)
Comment (smessmer) [reply caffe2/operators/layer_norm_op.cc]: I think being explicit is always better and more readable. Relying on input size can be confusing. But we should probably add asserts to make sure we're in one of the valid states.

File: caffe2/utils/math/reduce.cc (hunk 5 above)
Comment (BIT-silence) [reply caffe2/utils/math/reduce.cc]: Just in case some use case in PyTorch that the model is training in double.

File: caffe2/operators/layer_norm_op.cc (hunk 2 above)
Comment (BIT-silence) [reply caffe2/operators/layer_norm_op.cc]: It will help to determine whether to compute gradients for gamma and beta.

**Review by houseroad — APPROVED** [review]
Looks good to me.
